p2pclient/dht.go

```diff
@@ -0,0 +1,354 @@
+package p2pclient
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+
+	ggio "github.com/gogo/protobuf/io"
+	cid "github.com/ipfs/go-cid"
+	crypto "github.com/libp2p/go-libp2p-crypto"
+	pb "github.com/libp2p/go-libp2p-daemon/pb"
+	peer "github.com/libp2p/go-libp2p-peer"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// PeerInfo wraps the PeerInfo message from our protobuf with richer types.
+type PeerInfo struct {
+	// Id is the peer's ID
+	ID peer.ID
+	// Addrs are the peer's listen addresses.
+	Addrs []ma.Multiaddr
+}
+
+func convertPbPeerInfo(pbi *pb.PeerInfo) (PeerInfo, error) {
+	if pbi == nil {
+		return PeerInfo{}, errors.New("null peerinfo")
+	}
+
+	id, err := peer.IDFromBytes(pbi.GetId())
+	if err != nil {
+		return PeerInfo{}, err
+	}
+
+	addrs := make([]ma.Multiaddr, 0, len(pbi.Addrs))
+	for _, addrbytes := range pbi.Addrs {
+		addr, err := ma.NewMultiaddrBytes(addrbytes)
+		if err != nil {
+			log.Errorf("error parsing multiaddr in peerinfo: %s", err)
+			continue
+		}
+		addrs = append(addrs, addr)
+	}
+
+	pi := PeerInfo{
+		ID:    id,
+		Addrs: addrs,
+	}
+
+	return pi, nil
+}
+
+func newDHTReq(req *pb.DHTRequest) *pb.Request {
+	return &pb.Request{
+		Type: pb.Request_DHT.Enum(),
+		Dht:  req,
+	}
+}
+
+func readDhtResponseStream(ctx context.Context, control net.Conn) (<-chan *pb.DHTResponse, error) {
+	r := ggio.NewDelimitedReader(control, MessageSizeMax)
+	msg := &pb.Response{}
+	if err := r.ReadMsg(msg); err != nil {
+		return nil, err
+	}
+	if msg.GetType() != pb.Response_OK {
+		return nil, errors.New(msg.GetError().GetMsg())
+	}
+	if msg.Dht.GetType() != pb.DHTResponse_BEGIN {
+		return nil, fmt.Errorf("expected a stream BEGIN message but got %s", msg.Dht.GetType().String())
+	}
+
+	out := make(chan *pb.DHTResponse)
+	go func() {
+		defer close(out)
+		defer control.Close()
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				msg := &pb.DHTResponse{}
+				if err := r.ReadMsg(msg); err != nil {
+					log.Errorf("reading daemon response: %s", err)
+					return
+				}
+
+				if msg.GetType() == pb.DHTResponse_END {
+					return
+				}
+
+				out <- msg
+			}
+		}
+	}()
+
+	return out, nil
+}
+
+func (c *Client) request(req *pb.Request) (*pb.DHTResponse, error) {
+	control, err := c.newControlConn()
+	if err != nil {
+		return nil, err
+	}
+	defer control.Close()
+
+	w := ggio.NewDelimitedWriter(control)
+	if err = w.WriteMsg(req); err != nil {
+		return nil, err
+	}
+
+	r := ggio.NewDelimitedReader(control, MessageSizeMax)
+	msg := &pb.Response{}
+	if err = r.ReadMsg(msg); err != nil {
+		return nil, err
+	}
+
+	if msg.GetType() == pb.Response_ERROR {
+		err := fmt.Errorf("error from daemon in %s response: %s", req.GetType().String(), msg.GetError())
+		log.Errorf(err.Error())
+		return nil, err
+	}
+
+	return msg.GetDht(), nil
+}
+
+func (c *Client) requestNonNil(req *pb.Request) (*pb.DHTResponse, error) {
+	resp, err := c.request(req)
+	if err != nil && resp == nil {
```

Comment:
this is wrong, it will eat the error! It should be `if err == nil && resp == nil`.